### DISCUSSION TITLE
Porting the 2d Hittest Node from vvvv

### DISCUSSION THREAD

**Discussion by wirmachenbunt** [cables-gl/cables_docs#316] [Answers & Questions] [answered]
i'd like to port a rectangle hittest from vvvv. What i need in cables js code is:

1. the transform matrix from the transform op
2. and invert the matrix
3. and multiply it by a Vector2 containing the Mouse Coordinates

in C# the stripped down code looks like

`
Vector2D v = new Vector2D(´mouseX, mouseY);

                   
                        Matrix4x4 trobject;
                        this.FPinInTransform.GetMatrix(j, out trobject);
                        Vector2D trv = ((!trobject) * v).xy;

                        if (trv > -0.5 && trv < 0.5)
                        {
                            we have a hit.
                        }
`

in cables, is mat4.copy(m, cgl.mMatrix); getting the tranform op matrix ?
what would be the command to invert it ?
can i just multiply it by a Vector2 ?

it is actually the line  Vector2D trv = ((!trobject) * v).xy; which i don't know how to write in JS

https://github.com/vvvv/vvvv-sdk/blob/develop/vvvv45/addonpack/src/nodes/plugins/2d/HitTests/Nodes/QuadHitTestNode.cs

**Comment by wirmachenbunt** [comment]
one step further i guess
mat4.copy(m, cgl.mMatrix); looks like the right matrix
mat4.invert(invMat,m); looks like invert

**Comment by AMajesticSeaFlapFlap** [comment, marked as answer]
`mat4.invert(target, source)`
so probably
`var mmat=mat4.create(); mat4.invert(mmat, cgl.mMatrix);`

**Reply by AMajesticSeaFlapFlap** [reply]
to transform a vector with this matrix use `vec4.transformMat4(target, source, matrix)`

**Reply by wirmachenbunt** [reply]
mmm, this seems to take scale and rotation into account but not translation.
its almost working, i can rotate and scale the rectangle and it catches the hit but when i translate the rectangle, it still only does the hittest in the middle of the scene.

**Reply by wirmachenbunt** [reply]
ha...vec4 z and w need to be 1 to get it working

**Comment by wirmachenbunt** [comment]
![hittest](https://user-images.githubusercontent.com/1138673/184996533-943091e9-31f3-440a-9e25-e2235a2c5fa1.gif)
